### PR TITLE
fix: #5 Check for deleted files on startup

### PR DESCRIPTION
**Pull request by Julusian**
Checks for any media in the database that either don't reside in the current mediaPath, or do not exist on disk

### CODE REVIEW

**Comment by ronag** [comment]
This will not work very well when there are thousands of files registered in the db...

**Comment by ronag** [comment]
You should page the database like 100 elements at a time.

**Comment by Julusian** [comment]
Will that be a problem? If so then surely it will for /cls too? I am generating a db with 1000s of items to check how it performs anyway (which is taking quite a while). 
With 1000 items it is taking <1s, it is hard to tell how accurate that is as pouchdb is taking ~20s at startup to respond to queries, even without my changes

I realise that Promise.all is probably not the best option here, so am looking at that.

**Comment by Julusian** [comment]
I've made some small tweaks, but after testing it with a db of 7900 dead items it takes my machine 500ms-1s to load the data from the db and ~15s to check their status and remove them from the db. When everything can be found, the check time becomes <100ms.
Do you still want it to page the database?

**Comment by ronag** [comment]
```js
async function cleanDeleted () {
  let limit = 256
  let skip = 0
  while (true) {
    const { rows } = await db.allDocs({
      include_docs: true,
      skip,
      limit
    })
    logger.info('Checking for dead media')
    await Promise.all(rows.map(async ({ doc }) => {
      try {
        if (doc.mediaPath.indexOf(config.scanner.paths) === 0) {
          try {
            const stat = await statAsync(doc.mediaPath)
            if (stat.isFile()) {
              return
            }
          } catch (e) {
            // File not found
          }
        }
        await db.remove(doc)
        logger.info({ doc }, 'Removed')
      } catch (err) {
        logger.error({ err, doc })
      }
    }))
    if (rows.length < limit) {
      break
    }
  }

  logger.info('Checked for dead media')
}
```

**Comment by Julusian** [comment]
OK. This is now paging the results.

I had to make a couple of changes from what you did.
skip is only incremented if a file was valid, otherwise it wasnt processing everything.
It no longer logs the whole doc, as that is by far the slowest part of this process

**Comment by ronag** [comment]
The skip change is weird. It looks wrong. I don’t see what was wrong in my example? You want to page through the db.

**Comment by Julusian** [comment]
Paging like that would work if it was working with ids, not positions in the dataset. But because we are using positions, if we delete anything, then the position of anything after it changes to fill the gap.
An example:
data: [1,2,3,4,5] skip: 0 page: [1,2]
we delete 2, so the next iteration:
data: [1,3,4,5] with skip:2 page: [4,5]
but we didnt ever get 3

Thinking about this more, this strategy still has some potential race conditions if the user deletes some files while the check is running, as that could make out skip value no longer point to where we think it does

**Comment by ronag** [comment]
@Julusian: Don't delete anything in the loop. Keep an array of all docs to be deleted and do the delete in a batch (`bulkDocs`) at the end.

**Comment by ronag** [comment]
```js
  async function cleanDeleted () {
    logger.info('Checking for dead media')
    const deleted = []

    const limit = 256
    for (let skip = 0; true; skip += limit) {
      const { rows } = await db.allDocs({
        include_docs: true,
        skip,
        limit
      })
      await Promise.all(rows.map(async ({ doc }) => {
        try {
          if (doc.mediaPath.indexOf(config.scanner.paths) === 0) {
            try {
              const stat = await statAsync(doc.mediaPath)
              if (stat.isFile()) {
                return
              }
            } catch (e) {
              // File not found
            }
          }
          deleted.push({
            _id: doc._id,
            _rev: doc._rev,
            _deleted: true
          })
          logger.info({ id: doc.id, path: doc.mediaPath }, 'Removed')
        } catch (err) {
          logger.error({ err, doc })
        }
      }))
      if (rows.length < limit) {
        break
      }
    }

    await db.bulkDocs(deleted)

    logger.info('Checked for dead media')
  }
```

**Comment by ronag** [comment]
Why do you do the `doc.mediaPath.indexOf(config.scanner.paths) === 0` check?

**Comment by Julusian** [comment]
The docs say it is possible to page by id, so I would rather do that to be safer against the database changing from the scanner running its updates.

To ensure media reside inside the path. Otherwise if I change the path in casparcg.config, I get left with a bunch of inaccessible media becuase its outside the path, but it passes the other checks as it does still exist.

File: src/scanner.js [hunk 1]
```diff
@@ -45,6 +45,54 @@ module.exports = function ({ config, db, logger }) {
...
+          }
+
+          deleted.push({
+            _id: doc._id,
+            _rev: doc._rev,
+            _deleted: true
+          })
+        } catch (err) {
+          logger.error({ err, doc })
+          startkey = doc._id
```
Comment (ronag) [inline src/scanner.js]: Why set `startkey` here?

File: src/scanner.js [hunk 2]
```diff
@@ -45,6 +45,54 @@ module.exports = function ({ config, db, logger }) {
...
+        }
+      }))
```
Comment (ronag) [inline src/scanner.js]: If you page using `startkey` you might as well put `bulkDocs` in here to reduce memory usage from `deleted`.

File: src/scanner.js (hunk 1 above)
Comment (Julusian) [reply src/scanner.js]: That shouldn't be there. I missed it when changing when I updated startKey

**Pull request merged**